### PR TITLE
feat: init matrix coordinate

### PR DESCRIPTION
**Pull request by Ovilia**
## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [x] new feature
- [ ] others

## Document Info

One of the following should be checked.

- [ ] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

N.A.

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

File: src/coord/matrix/Matrix.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+import { RectLike } from 'zrender/src/core/BoundingRect';
+import { SeriesModel } from '../../echarts.all';
```
Comment (plainheart) [inline src/coord/matrix/Matrix.ts:21]: Unexpected import from `echarts.all`. Should be:

```ts
import type SeriesModel from '../../model/Series';
```

File: src/component/matrix/MatrixView.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                        width: cellWidth,
+                        height: cellHeight
+                    },
+                    style: innerBackgroundStyle
+                }));
+            }
+        }
+
+        // Outer border
+        const backgroundStyle = matrixModel
```
Comment (plainheart) [inline src/component/matrix/MatrixView.ts:143]: These borders may overlap and cause a darker border color. I'm not sure if we should provide a similar ability like the CSS table property [`border-collapse`](https://developer.mozilla.org/docs/Web/CSS/border-collapse). It looks not easy to implement that feature.

And the chart seems rendering unexpectedly if I set the width of both the outer border and inner border to `0`.

```js
backgroundStyle: {
    borderWidth: 0
},
innerBackgroundStyle: {
    borderWidth: 0
}
```

**Review by plainheart — COMMENTED** [review]
The package.json may need to be updated as well. 🤔
https://github.com/apache/echarts/blob/ab3efd31510c52b7b2be447025e83adfcbbcfff3/package.json#L172

**Comment by admsev** [comment]
@gooroodev review

**Comment by gooroodev** [comment]
### Summary of Changes
1. **New Matrix Coordinate System:**
   - Added a new coordinate system called "matrix" which allows for hierarchical data representation in a grid-like format.
   - Implemented the necessary classes and methods to support the matrix coordinate system, including `Matrix`, `MatrixDim`, `MatrixModel`, and `MatrixView`.
   - Integrated the matrix coordinate system with existing components like heatmap and pie charts.

2. **Component Registration:**
   - Added the matrix component to the list of built-in components.
   - Registered the matrix coordinate system in `echarts.all.ts`.

3. **Heatmap and Pie Chart Integration:**
   - Modified `HeatmapView` to support rendering on the matrix coordinate system.
   - Updated `PieSeries` to handle matrix coordinate system data dimensions.

4. **Theming:**
   - Added default theming for the matrix coordinate system in the dark theme.

5. **Test Cases:**
   - Added HTML test files (`matrix.html` and `matrix_application.html`) to demonstrate and test the new matrix coordinate system.

### Issues, Bugs, or Typos
1. **Unused Imports:**
   - In `HeatmapView.ts`, the import `import type Element from 'zrender/src/Element';` is not used.

2. **Console Warning:**
   - The `containPoint` method in `Matrix.ts` currently logs a warning and returns false. This should be implemented or removed.

### Proposed Code Improvements
1. **Remove Unused Imports:**
   ```typescript
   // In src/chart/heatmap/HeatmapView.ts
   // Remove this line
   import type Element from 'zrender/src/Element';
   ```

2. **Implement `containPoint` Method:**
   ```typescript
   // In src/coord/matrix/Matrix.ts
   containPoint(point: number[]): boolean {
       const rect = this._rect;
       return point[0] >= rect.x && point[0] <= rect.x + rect.width
           && point[1] >= rect.y && point[1] <= rect.y + rect.height;
   }
   ```

### General Review of Code Quality and Style
- **Code Quality:**
  - The code is well-structured and follows consistent naming conventions.
  - The use of TypeScript's type system is effective and helps in maintaining type safety.
  - The code is modular, with clear separation of concerns between different components and functionalities.

- **Style:**
  - The code adheres to common coding standards and practices.
  - Comments and documentation are present, making the code easier to understand.
  - The use of ES6 features such as destructuring and arrow functions enhances readability.

- **Testing:**
  - The addition of test cases in HTML files is a good practice to ensure the new functionality works as expected.
  - More comprehensive test cases could be added to cover edge cases and potential failure scenarios.

Overall, the pull request introduces a significant new feature with a well-thought-out implementation. The code quality is high, and the proposed improvements are minor and aimed at further enhancing the robustness and maintainability of the code.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)

File: src/component/matrix/MatrixView.ts (hunk 2 above)
Comment (Ovilia) [reply src/component/matrix/MatrixView.ts:143]: The border won't be darker unless the opacity of borderColors are not 1. I personally think providing a `border-collapse` like mechanism is over-kill because then, we should also implement other features like border widths and colors being different values at four sides. This should bring much work to implement it robustly. The reason why I choose to render rectangles with background and border is that it would be more difficult to think which option controls which border line.

The image you provided is as expected. You should add `x.itemStyle.borderWidth: 0` and `y.itemStyle.borderWidth: 0` if you want to remove those rest borders.

**Comment by 100pah** [comment]
During the CR of this PR, I found that there are features can be added or enhanced to make a more comprehensive coordinate system definition and implementation. Therefore I made a implement in https://github.com/apache/echarts/pull/21005 , and seek to merge to this PR. And all the comments I suggested in this CR has been fixed in that suplement PR.


NOTE: this PR requires https://github.com/ecomfe/zrender/pull/1122

#21005 made these changes:

## Support declaratively layout other coorSys/components/series based on matrix and calendar

Some abstraction has been make based on calendar and matrix - they are similar and represent a kind of coord sys that can convert data value to a "container(rect)", rather than only a "point".

Therefore, all of the other components can be layout on matrix/calendar. including:
- coord sys components, such as, grid/polar/geo/graph/parallel/singleAxis/...
- non-coordSys components, such as, legend/title/dataZoom/visualMap/toolbox/...
- series, such as, pie/funnel/map/radar 

The effect is:

image

image

image


bbb
(Responsive grid layout)

Previously we also made case that layout cartesian or pie on calendar, but that cases relies on `chart.convertToPixel` - not a declarative way and complicated.

This supplement PR introduced the uniform declarative way to do that:

```ts
option: {
    matrix: {...},
    grid: {
        coordinateSystem: 'matrix',
        coord: ['X2', 'Y6'], // coord on matrix.
    },
}
```
That is, the option `coordinateSystem` and `coord` (newly introduced) can be used universally to any components/series.

Additionally, there are some complicated case in `series.graph`/`series.map`: They can use a coord sys to lay out both each item in `series.data` or lay out the overall bounding rect. To distinguish the usage of the declared coord sys, the supplement PR introduce a new option `coordinateSystemUsage`: 
```ts
option: {
    matrix: {...},
    graph: {
        coordinateSystem: 'matrix',
        /**
         * - "data": Use it as "dataCoordSys", each data item is laid out based on a coord sys.
         * - "box": Use it as "boxCoordSys", the overall bounding rect or anchor point is calculated based on a coord sys.
         *   e.g.,
         *      grid rect (cartesian rect) is calculate based on matrix/calendar coord sys;
         *      pie center is calculated based on calendar/cartesian;
         *
         * The default value (if not declared in option `coordinateSystemUsage`):
         *  For series, be "data", since this is the most case and backward compatible.
         *  For non-series components, be "box", since "data" is not applicable.
         */
        coordinateSystemUsage: 'box',
        coord: ['X2', 'Y6'], // coord on matrix.
    },
}
```
In most cases, `coordinateSystemUsage` do not need to specified, since there is no ambiguity about the usage of a coord sys.

It's worth noting that this supplement PR only supports `matrix` and `calendar` to be a "box layout base" to other coord sys, but do not support such as layout `cartesian2d` on `geo` declaratively, and do not recursive dependencies, such as, layout a `matrix` on another `matrix`. They can be implemented thearetically, but it requires some more complicated logic, see the comment in `src/core/CoordinateSystem.ts` for details.


Regarding the implementation, previously almost all of the components can be uniformly layout by option `left/right/top/bottom/width/height`, and internally uniformly handled by `getLayoutRect` in `src/layout.ts`. This new impl adds a new uniform method `createBoxLayoutReference` to layout them based on the canvas or the declared coord sys.



## More sophisticated APIs

Along with this feature, a new echarts API `chart.convertToLayout` is introduced, providing the ability to return layout info (currently, `rect` or `contentRect`), rather than previously `chart.convertToPixel` can only return a point. And in custom series, a new renderItem param API `api.layout` is added corresponding to `chart.convertToLayout`, just as the previous custom series API `api.coord` corresponds to `chart.convertToPixel`.

And previous calendar and matrix `dataToRect` has been removed; Use `dataToLayout` instead.

And both `chart.convertToLayout` and originally existing API `chart.converFromPixel` and `chart.convertToPixel` have been added a parameter `opt` to convey some config info, such as `clamp`. Matix coord sys can have some more detailed `clamp` option, such as, clamp in body, in corner. 

Based on that APIs, some complex effect can be make. This is a test case for that APIs (`convertToLayout` and `convertFromPixel`:

aaaaa



## Support matrix col/row size option

Column/row size need to be able to specified.

The effect:
image


```ts
option = {
    matrix: {
        x: {
            data: [
                // The width of a specific column
                {value: 'A', size: 123},
            ],
            // level size for all levels (x header rows).
            levelSize: 55
            // level size for a specific level (a x header row).
            levels: [{levelSize: 10}],
        },
        y: {
            data: [
                // The height of a specific row
                {value: 'M', size: 123},
            ],
            // level size for all levels (y header columns).
            levelSize: 123
            // level size for a specific level (a y header column).
            levels: [null, null, {levelSize: 32}],
        }

    }
}
```

**PENDING: size naming** :
Currently `matrix.x` and `matrix.y` use the unified size option name `size` and `levelSize`.
But is it intuitive enough to understand?
Do we need to support `columnSize` in `matrix.x` and `rowSize` in `matrix.y`?

## Support matrix `mergeCells`
Cell merge is commonly used in table or spreadsheet.
See the effect:
image


The `matrix.x/y` has a tree hierarchy and naturally has cell span. The supplement PR preserve that definition and still do not allow specifying random cell merging.

But in body and corner, the supplement PR introduce the cell merging definitions:

```ts
option = {
    matrix: {
        x: {
            data: ['X5', 'X6', 'X7']
        },
        y: {
            data: ['Y90', 'Y91', 'Y92']
        },
        body: {
            data: [
                // Merge the rect defined by x range 'X5', 'X6' and y range 'Y90'
                {coord: [['X5', 'X6'], 'Y90'], mergeCells: true}, 
                // Merge the entire row
                {coord: [null, 'Y92'], mergeCells: true}, 
            ]
        }
    }
};
```


## Support "corner" (top-left area)
It's might be a challenge that how to design the option for users to customize the corner.
I think the corner can be defined as the same way as "body", that is, the corner can also be composed by cells, and share the `mergeCells` and style options formatter.

See the effect:

image


In the supplement PR, corner can be located by a pair of negative integers. See "locator of matrix" below.


## Clarify the locator of matrix
Suport locate a cell (either in body or in corner or in header(x/y dimension)).
Basically, a locator and is defined by `matrix.x/y.data`, and can be a string (such as, 'X1', 'Y1', ) or `OrdinalNumber`(i.e, -2, -1, 0, 1, 2, 3, ...)

Below is the details. 

TL;DR,

The rule is applied uniformly (such as, in `matrix.dataToPoint` and `matrix.dataToLayout` and `xxxComponent.coord`)

Suppose the matrix.x/y dimensions (header) are defined as:
```ts
matrix: {
    x: [{ value: 'Xa0', children: ['Xb0', 'Xb1'] }, 'Xa1'],
    y: [{ value: 'Ya0', children: ['Yb0', 'Yb1'] }],
}
```
```
-----------------------------------------
|       |       |     Xa0       |       |
|-------+-------+---------------|  Xa1  |
|cornerQ|cornerP|  Xb0  |  Xb1  |       |
|-------+-------+-------+-------+--------
|       |  Yb0  | bodyR | bodyS |       |
|  Ya0  |-------+-------+---------------|
|       |  Yb1  |       |     bodyT     |
|---------------|------------------------
```
 
+ "Locator number" (`MatrixXYLocator`):
    + The term `locator` refers to a integer number to locate cells on x or y direction.
    + Use the top-left cell of the body as the origin point (0, 0),
        + the non-negative locator indicates the right/bottom of the origin point;
        + the negative locator indicates the left/top of the origin point.
+ "Ordinal number" (`OrdinalNumber`):
    + This term follows the same meaning as that in category axis of cartesian. They are non-negative integer, designating each string `matrix.x.data[i].value`/`matrix.y.data[i].value`. 'Xb0', 'Xb2', 'Xa1', 'Xa0' are assigned with the ordinal numbers 0, 1, 2, 3.
    + For every leaf dimension cell, `OrdinalNumber` and `MatrixXYLocator` is the same.
+ A cell or pixel point or rect can be determined/located by a pair of `MatrixCoordValueOption`.
    + See also `MatrixBodyCornerCellOption['coord']`.
- The body cell `bodyS` above can be located by:
    - `coord: [1, 0]` (`MatrixXYLocator` or `OrdinalNumber`, which is a non-negative integer)
    - `coord: ['Xb1', 'Yb0']`
    - `coord: ['Xb1', 0]` (mix them)
- The corner cell `cornerQ` above can be located by:
    - `coord: [-2, -1]` (negative `MatrixXYLocator`)
    - But it is NOT supported to use `coord: ['Y1_0', 'X1_0']` (XY transposed form) here. It's mathematically sound, but may introduce confusion and unnecessary complexity (consider the 'Xa1' case), and corner locating is not frequently used.
- `mergeCells`: Body cells or corner cells can be merged, such as "bodyT" above, an input
    - The merging can be defined by: `matrix.data[i]: {coord: [['Xb1', 'Xa1'], 'Yb0'], mergeCells: true}`.
    - Input `['Xa1', 'Yb1']` to `dataToPoint` will get a point in the center of "bodyT".
    - Input `['Xa1', 'Yb1']` to `dataToLayout` will get a rect of the "bodyT".
- If inputing a non-leaf dimension cell to locate, such as `['Xa0', 'Yb0']`,
    - it returns only according to the center of the dimension cells, regardless of the body span. (therefore, the result can be on the boundary of two body cells.) And the oridinal number assigned to 'Xa0' is 3, thus input `[3, 'Yb0']` get the some result.
- The dimension (header) cell can be located by negative `MatrixXYLocator`. For example:
    - The center of the node 'Ya0' can be located by `[-2, 'Ya0']`.


## Fix and refactor the matrix style options
Considered their inheritance and z-order and flexibility.

image


Considering the flexibility of the style configuration, the style related options are sort to:

```ts
option = {
    matrix: {
        x: {
            // Two levels of inheritance itemStyle and label.
            itemStyle: {...},
            label: {...},
            // Padding for the inner content (label / series / other coord sys).
            padding?: number | number[];
            cursor?: string;
            // By default, auto decide whether to be silent, considering tooltip.
            silent?: boolean | NullUndefined;
            // Used when style conflict - especially for thick border style.
            z2?: number;

            dividerLineStyle: {...} // Special for the divider line.

            data: [
                {value: ..., itemStyle: {...}, label: {...}, padding, cursor, silent, z2},
                {value: ..., itemStyle: {...}, label: {...}, padding, cursor, silent, z2},
            ],
        },
        y: { /* The same as x */},
        body: { /* The same as x */},
        corner: { /* The same as x */},
        backgroundStyle: {...} // Special for the overall background.
    }
}
```

## Support that auto collect matrix dimension data from series
This behaves the same with category axis, that is:
```ts
option = {
    matrix: {} // no x y defined,
    dataset: {
        source: [
            ['2021-02-01', 'amount', 1212, 'file', 2321, 'Q', 1412],
            ['2021-02-02', 'amount', 7181, 'file', 2114, 'Q', 1402],
            ['2021-02-03', 'amount', 2763, 'file', 4212, 'Q', 8172],
            ['2021-02-04', 'amount', 6122, 'file', 2942, 'Q', 6121],
            ['2021-02-05', 'amount', 4221, 'file', 3411, 'Q', 1987],
            ['2021-02-06', 'amount', 7221, 'file', 5121, 'Q', 1303],
        ]
    },
    series: [{
        type: 'scatter',
        coordinateSystem: 'matrix',
        encode: {x: 1, y: 0, label: 2},
    }, {
        type: 'scatter',
        coordinateSystem: 'matrix',
        symbol: 'rect',
        encode: {x: 3, y: 0, label: 4},
    }, {
        type: 'scatter',
        coordinateSystem: 'matrix',
        encode: {x: 5, y: 0, label: 6},
    }],
};
```

## Cell text (in header and body and corner)
+ Support overflow settings for long header text.
    + `label.overflow: 'truncate' | 'break' | | 'breakAll' | 'none' | undefined | null`
    + `label.ellipsis: '...'`
    + `label.lineOverflow: 'truncate' | undefined | null` (this option is available to all `label` options in echarts, previously not)
+ Support tooltip for truncated text (truncated is the default option).
    ```js
    // Enable tooltip on matix.
    option = {
        tooltip: {},
        matrix: {
            tooltip: {show: true},
            // ...
        }
    }
    // By default tooltip is available only on text.
    // If intending to enable tooltip on the entire cell, 
    //  specify `matrix.x/y/body/corner.silent: false` or `matrix.x/y/body/corner.data[i].silent: false`
    ```
+ Support other alignment / position settings for text, rather than only 'center'. 
    - e.g., `matrix.x.data[i].label.position = 'innerRight'`.

## Others

+ Subpixel optimization for rect line is added.
+ Edge cases has been tested.
+ Uniform the query of coord sys in matrix, calendar, geo, make them consistent with other components: both `xxxIndex` and `xxxId` are available.
+ Optimize the query from the original PR, use hash map.
+ TS type enhanced.
+ Naming: in the context of 2d-table, the term "inner", "innerCell" seem not common. Changed to be "body cell". In html, th td represent "header" and "data"; in other context, "header"-"body" is commonly used.
+ visualMap support reference series by both `seriesId` and `seriesIndex`
    + Previous behavior: can only by `seriesIndex`, and if `seriesIndex == null`, reference all series.
    + Behavior: if `seriesIndex == null && seriesId == null`, reference all series.
+ geo coord sys can be referenced by both `series.geoIndex` and `series.geoId`
    + Previous behavior: can only by `geoId`.

**Comment by plainheart** [comment]
I noticed this PR is based on the 'next' branch. Should we change it to the 'v6' branch?

**Comment by 100pah** [comment]
# Roaming Infrastructure Upgrade

## Support `preserveAspect`

Support `preserveAspect` `preserveAspectAlign` `preserveAspectVerticalAlign` on `series.map`, `geo`, `series.graph` in box layout (lay out by `left`/`right`/`top`/`bottom`/`width`/`height`).

xx1

The new introduced echarts option is:
```ts
    // Suppose a "viewport" is decided by `left`/`right`/`top`/`bottom`/`width`/`height`,
    // - null/undefined/false (default): aspect ratio will not be preserved, but stretched to fill
    //   the "viewport".
    // - 'contain'/true: The aspect ratio is preserved; the viewRect is contained by the "viewport",
    //   and scaled up as much as possible to meet the "viewport".
    // - 'cover': The aspect ratio is preserved; the viewRect covers the "viewport", and scaled down
    //   as much as possible to meet the "viewport".
    preserveAspect?: boolean | 'contain' | 'cover';
    // By default 'center'
    preserveAspectAlign?: 'left' | 'right' | 'center';
    // By default 'middle'
    preserveAspectVerticalAlign?: 'top' | 'bottom' | 'middle';
```
For example:
```js
chart.setOption({
    geo: [{
        left: 10,
        right: 10,
        top: 10,
        bottom: 10,
        preserveAspect: true,
        // ...
    }]
});
// or 
chart.setOption({
    series: [{
        type: 'graph',
        left: 10,
        right: 10,
        top: 10,
        bottom: 10,
        preserveAspect: true,
        // ...
    }]
});
```

## Support `clip`

Support `clip` on `geo` component and `series.map`.

That is, if `clip: true`, hide the part that overflows the given layout rect defined by `left`/`right`/`top`/`bottom`/`width`/`height`.

xx2

For example:
```js
chart.setOption({
    geo: [{
        clip: true, // default `false`
        // These props also defines the clip area.
        left: 20, 
        top: 50, 
        right: 50, 
        bottom: 50, 
        // ...
    }]
});
// or
chart.setOption({
    series: [{
        type: 'map',
        clip: true, // default `false`
        // These props also defines the clip area.
        left: 20, 
        top: 50, 
        right: 50, 
        bottom: 50, 
        // ...
    }]
});
```


## Support cursor style change on roaming.

Support `roamTigger: 'global' | 'selfRect'` to switch roam area to global on only self bounding rect on `geo`, `series.map`, `series.graph`, `series.tree`, `series.sankey`, `series.treemap`.

Before:
bb2

After:

bb1

Support cursor style change when hovering the roaming enabled area to hint users.

## Support `roamTrigger: 'global' | 'selfRect'`

xx3

```ts
    /**
     * Hover over an area where roaming is triggered.
     * - if `null`/`undefined`/`'selfRect'`, the trigger area is
     *   the intersection of "self bounding rect" and "clipping rect (if any)".
     * - if 'global', the trigger area is
     *   the intersection of "the entire canvas" and "clipping rect (if any)".
     * NOTE:
     *  The clipping rect, which can be enabled by `clip: true`, is typically the layout rect.
     *  The layout rect is typically determined by option `left`/`right`/`top`/`bottom`/`width`/`height`, some
     *  components/series, such as `geo` and `series.map` can also be determined by `layoutCenter`/`layoutSize`,
     *  and may modified by `preserveAspect`.
     */
    roamTrigger?: 'global' | 'selfRect' | NullUndefined
```
For example: 
```js
chart.setOption({
    geo: {
        roamTrigger: 'global', // by default 'selfRect' in geo component.
        // ...
    }
});
```


## Fix the percent base of `center` for all `coord/View` coord sys based components/series

"View coord sys based components/series" is `geo`/`series.map`/`series.graph`/`series.tree`/`series.sankey` 
Percent base of the `center` option is like:
```js
chart.setOption({
    geo: {
        center: ['40%', '50%'],
        // ...
    }
});
```
Notice that historically the definition of this `center` option has always been irrelevant to some other series center like `series[type='pie'].center` - this center is a point on the raw data coord sys, rather than canvas viewport, and the unit is not necessarily pixel. For example, in `geo` component, the `center` option is `[longitude, latitude]`. This `center` option  means that moving that center point to the center of the viewport.

#16904 introduced percentage string here, such as '33%'. But it was based on canvas width/height, which is not reasonable - the unit may incorrect, and it is unpredictable if the raw data coords are not calculated based on the current canvas rect. Therefore, the percentage value is changed to based on the raw data bounding rect since v6. Under this definition, users can use '0%' to move the top/left of the content to the center of the viewport, and use '100%' to move the bottom/right of the content to the center of the viewport.

This is a breaking change. To restore the previous behavior in v5:
```js
chart.setOption({
    legacyViewCoordSysCenterBase: true,
    // ... 
});
```


## Others

## Enhance the behavior for roaming area overlapping 
The upper one has higher precedence.

### Fix some `scaleLimit` missing

### Fix center and zoom option does not work in series.sankey.
And no relevant test case. (the roam is introduced in impl in #20321, due v6).

### Some refactor and clarification of `RoamController` and `roamHelper`

**Comment by 100pah** [comment]
## Thumbnail

The final refactor of Thumbnail is make here.
The previous thumbnail related PR:
https://github.com/apache/echarts/pull/17471
https://github.com/apache/echarts/pull/18039


Thumbnail is finally designed as a component.
New echarts component option is introduced:
```js
const option = {
    series: {
        type: 'graph',
        // ...
    },
    thumbnail: {
        // ...
        seriesId: 'xxx', // Optional, use the first series.graph by default.
        seriesIndex: 0, // Optional, use the first series.graph by default.
    }
}
```
Currently thumbnail only support series.graph. 
More other base map, such as geo, series.tree, series.sankey, can be supported based on the current data  structure.

The thumbnail options:

```ts
{
    type?: string;

    id?: OptionId;
    name?: OptionName;

    z?: number;
    zlevel?: number;

    coordinateSystem?: string
    coordinateSystemUsage?: CoordinateSystemUsageOption
    coord?: CoordinateSystemDataCoord
    width?: PositionSizeOption;
    height?: PositionSizeOption;
    top?: PositionSizeOption;
    right?: PositionSizeOption;
    bottom?: PositionSizeOption;
    left?: PositionSizeOption;

    show?: boolean

    itemStyle?: ItemStyleOption
    windowStyle?: ItemStyleOption

    seriesIndex?: number | number[]
    seriesId?: string | string[]
}
```

The example is `test/graph-thumbnail.html`.

The minimal echarts option to enable thumbnail on graph: 
```js
chart.setOption({
    series: {
        type: 'graph',
        data: [{
            itemStyle: null,
            name: 'DRD2',
            value: 40,
            symbolSize: 40
        }, {
            itemStyle: null,
            name: 'ADORA2A',
            value: 0,
            symbolSize: 20
        }],
        links: [],
    },
    thumbnail: {}, // declare a thumbnail component, use the first series.graph by default.
});
```
